### ISSUE TITLE
Tablet Input Broken on Wayland

### ISSUE THREAD

**Issue by andersonjwan** [blackhole89/notekit#44]
Hello, so I have just been recently using Notekit; however, I have ran across a bug where the input system does not properly draw from the tablet when using Wayland. I tested on Xorg, and the input works perfectly, but I am wondering why Wayland seems to have an issue.

It seems that a stroke object is created, but the actually drawing does not work with the input from the tablet leaving a blank object. Using a mouse input and draw tool seems to work, however.

Let me know if you need any further details to look into and/or resolve the issue. Thanks!

**Comment by blackhole89** [comment]
Unfortunately, I don't have access to any Wayland-based system to test it on myself. I think it's most likely that there is some issue with the pressure detection. Does the drawing that is created when painting with the tablet reflect the size of the painting correctly? (i.e. is it an empty box corresponding in size to what you drew?) Also, when drawing with the tablet, is there any output on the command line of the form `old: <number>, new: <number>`?

**Comment by andersonjwan** [comment]
Yes, there seems to be a bounding box around the "blank" drawing. There seems to be output on the terminal, so it's being recognized, just the drawing itself isn't shown. I have attached a screenshot below to better help debug it.

![screenshot_002](https://user-images.githubusercontent.com/52432873/91105312-9f0f1b80-e624-11ea-806f-d3e22e49db92.png)

`== This is notekit, built at Mon Apr 27 08:48:48 2020. ==
Detected paths:
Config: /home/andersonjwan/.config/notekit
Active notes path: /home/andersonjwan/.local/share/notekit
Default notes path: /home/andersonjwan/.local/share/notekit
Resource path: /usr/share//notekit

color1
autosave
old: 11, new: 6
old: 12, new: 6
old: 11, new: 5
old: 6, new: 4
old: 8, new: 3
old: 3, new: 3
old: 5, new: 4
old: 16, new: 11
old: 7, new: 4
old: 7, new: 3
autosave`

**Comment by sp1ritCS** [comment]
Same issue on wayland

this is what is saved in the file:
https://bin.privacytools.io/?a4d150caab932be1#Ug6HCp6xjghdiLF/Ah4dUsJBNjhHS3gE5/ENCQ4lgBM=

**Comment by blackhole89** [comment]
Thanks for the file. As I suspected, the pressure for all the strokes in the wayland drawing is recorded as 0.

I've tried to reproduce this, but Weston for me straight up does not recognise my wacom digitiser, and I had no luck installing any other Wayland compositor because of dependency hell in my Frankendebian setup. What sort of system did you produce the file linked above in?

(I can try to hack in some kind of workaround for systems where the device reports having a "pressure" axis but always reads 0, but the best thing I could do would be pinning it to always have constant pressure, and it would not be a particularly good solution either way.)

**Comment by blackhole89** [comment]
If anyone with a Wayland setup with digitiser support wants to take a stab at this, the relevant logic is pretty compact (found in `CNotebook::ReadPressure` in `notebook.cpp`). The key is the part that reads

```cpp
    gdk_device_get_state(d,w,axes,NULL); 
    if(gdk_device_get_axis(d,axes,GDK_AXIS_PRESSURE,&r))
        return r;
```
- the expected behaviour would be that this returns `false` whenever GDK_AXIS_PRESSURE is not available (see [documentation](https://developer.gnome.org/gdk3/stable/GdkDevice.html#gdk-device-get-axis)), but in all likelihood that's not happening here.

**Comment by sp1ritCS** [comment]
> What sort of system did you produce the file linked above in?

openSUSE Tumbleweed, GNOME 

> The expected behaviour would be that this returns false whenever GDK_AXIS_PRESSURE is not available (see documentation), but in all likelihood that's not happening here.

Well, It should be available (also returns true), but it writes 0.0 to r

> If anyone with a Wayland setup with digitiser support wants to take a stab at this

I've tried, but I'm not a C Dev, so I had no Idea what I was doing; I found [this function on the GdkDevice Docs](https://developer.gnome.org/gdk3/stable/GdkDevice.html#gdk_device_get_axis_value)
```
gdk_device_get_axis_value(d, axes, gdk_atom_intern_static_string ("Abs Pressure"), &r) -> bool
```

But this time it returns false (dunno if it does this on X as well)

---

I don't know if it is possible for you to setup qemu/kvm/libvirt on your "Frankendebian setup" but if you do, you could pass through the usb of the tablet and run wayland in there.

If I can help you with anything let me know.
